CRM/Core/Payment/Dummy.php

```diff
@@ -137,6 +137,38 @@ protected function supportsLiveMode() {
     return TRUE;
   }
 
+  /**
+   * Does this payment processor support refund?
+   *
+   * @return bool
+   */
+  public function supportsRefund() {
+    return TRUE;
+  }
+
+  /**
+   * Submit a refund payment using Advanced Integration Method.
+   *
+   * @param array $params
+   *   Assoc array of input parameters for this transaction.
+   *
+   * @return array
+   *   the result in a nice formatted array (or an error object)
+   */
+  public function doRefundPayment(&$params) {
+    if (empty($params['trxn_id'])) {
+      $error = new CRM_Core_Error(ts('Refund failed'));
+      return $error;
+    }
+    else {
+      $params['payment_status_id'] = CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'contribution_status_id', 'Refunded');
+      $params['trxn_date'] = date('YmdHis');
+    }
+    $params['gross_amount'] = $params['amount'];
+
```

Comment:
This code is for dummy processor to mimic refund behavior for testing purposes.

Reply:
Hmm - what are we mimicing? This implies we expect those variables returned?

Reply:
This is the first time `gross_amount` is used in CiviCRM as far as I know, everywhere else is `total_amount` though we do have inconsistencies.  So I think either returning `total_amount` or making a design decision that we'd like to standardise on `gross_amount` throughout CiviCRM in the future.

Reply:
I don't think we should be returning anything from the Dummy processor unless we have a clear reason to - easier not to add stuff than to figure out whether it's actually being used & remove later

Reply:
@mattwire no this is not the first time. If you grep ```gross_amount``` under CRM/Core/Payment/ you will the find the occurrences in other Payment Processors. 

@eileenmcnaughton while testing with Moneris refund it returns a new trxn_id and other information (see [here](https://github.com/adixon/ca.civicrm.moneris/compare/master...JMAConsulting:refund-master?expand=1#diff-61e08438be5781d5849a1554cae3f01eR379)) So in order to do similar to Dummy I thought of setting at least status, amount(which would be same though), trxn_date. We can drop ```return @params``` here but shoudn't we setting the $params atleast?

Reply:
Yeah I think a) we should EITHER receive $params by reference OR return $params - doing both seems silly. 

I think it's fine to pass stuff back but I'd rather add to signature carefully & slowing. In general the existing functions have too many poorly-documented parameters that are not tested & may or may not do things

So we have 
- trxn_id - which makes sense as it would be the trxn_id of the new (refund) payment
- gross_amount - there are cases where this is being used already although probably it shouldn't be - if we return fee_amount then the calling function will be able to calculate other values & we won't have some processors doing things & others not
txn_date - didn't we pass that in? so we already know it
- payment_status_id  - so I guess failed is an exception & no exception is succeeded. 

But I guess there is another scenario 'pending' - I don't know if any of them would do this but it's a whole new scenario if we are going to support this

Reply:
I am in support of passing @params by reference. 

Re gross_amount good point, need to put some thoughts about fee_amount wether it would be worth considering.

Re txn_date hmm yes, atleast for dummy I will drop it but for other PP it might return a refund date in response. 

Re payment_status_id my idea is to, do nothing if there is an exception and it will halt any further operation. But if its a success lets set the Refunded status and it will be reflected in financial trxn. 

As a separate note I think how Civi should work is if a payment is entirely refunded then the contribution status should be set to Refunded or if part of the payment is refunded then the reverse financial trxn should have 'Refunded' status and its contribution as 'Completed' status?

Reply:
@monishdeb so what statuses could payment_status_id have - if a success will ALWAYS set to 'Refunded' then that isn't a job for the processor code (just a place to make mistakes) - if there are mutliple possibilitied then....?

Regarding Contribution status - that is out of scope for this to manage but you mean if the *contribution*, rather than payment, is fully refunded I think. Hopefully by the time we've finished fixing up the payment.create function that will have incorporated the handling from the payment form which hopefully (a lot of hope) already appropriately handles contribution status....

Reply:
@eileenmcnaughton yeah you are right. Looking down the code, its already handling the status to 'Refunded' so I shouldn't set it here, also not the case that entire core code is depending on that status. Will remove this from processer code.